src/v/storage/segment.h

```diff
@@ -40,27 +40,75 @@ struct segment_closed_exception final : std::exception {
 class segment {
 public:
     using generation_id = named_type<uint64_t, struct segment_gen_tag>;
-    struct offset_tracker {
+    class offset_tracker {
+    public:
+        using committed_offset_t
+          = named_type<model::offset, struct committed_offset_tag>;
+        using stable_offset_t
+          = named_type<model::offset, struct stable_offset_tag>;
+        using dirty_offset_t
+          = named_type<model::offset, struct dirty_offset_tag>;
+
         offset_tracker(model::term_id t, model::offset base)
-          : term(t)
-          , base_offset(base)
-          , committed_offset(model::prev_offset(base))
-          , dirty_offset(model::prev_offset(base))
-          , stable_offset(model::prev_offset(base)) {}
-        model::term_id term;
-        model::offset base_offset;
+          : _term(t)
+          , _base_offset(base)
+          , _committed_offset(model::prev_offset(base))
+          , _stable_offset(model::prev_offset(base))
+          , _dirty_offset(model::prev_offset(base)) {}
+
+        template<typename... Ts>
+        void set_offsets(Ts... ts) {
```

Comment:
Neat & clever! 👍

Reply:
Shall we also have a mean of enforcing that offsets don't move backwards? Unless it is a call to truncate?